### PR TITLE
chore: Make e2e tests runnable against remote cluster

### PR DESCRIPTION
**Pull request by jannfis**
This allows for running the end-to-end test suite against a real workload, deployed to a Kubernetes cluster.

Refer to the `README.md` for a more detailed explanation.

Signed-off-by: jannfis <jann@mistrust.net>

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by sbose78** [comment]
Thank you!

File: test/e2e/fixture/certs/certs.go [hunk 1]
```diff
@@ -14,31 +14,46 @@ import (
 func AddCustomCACert() {
 	caCertPath, err := filepath.Abs("../fixture/certs/argocd-test-ca.crt")
 	errors.CheckError(err)
-	args := []string{"cert", "add-tls", "localhost", "--from", caCertPath}
-	errors.FailOnErr(fixture.RunCli(args...))
-	args = []string{"cert", "add-tls", "127.0.0.1", "--from", caCertPath}
-	errors.FailOnErr(fixture.RunCli(args...))
+	if fixture.IsLocal() {
+		args := []string{"cert", "add-tls", "localhost", "--from", caCertPath}
```
Comment (sbose78) [inline test/e2e/fixture/certs/certs.go:18]: Could you add a comment please ? Feels a little obscure for someone not familiar with the code :D

File: test/e2e/fixture/certs/certs.go [hunk 2]
```diff
@@ -14,31 +14,46 @@ import (
...
+// an optional path to the known hosts file, instead of using the default one.
 func AddCustomSSHKnownHostsKeys() {
 	source := os.Getenv("ARGOCD_E2E_SSH_KNOWN_HOSTS")
 	if source == "" {
 		source = "../fixture/testrepos/ssh_known_hosts"
 	}
 	knownHostsPath, err := filepath.Abs(source)
 	errors.CheckError(err)
-	args := []string{"cert", "add-ssh", "--batch", "--from", knownHostsPath}
+	args := []string{"cert", "add-ssh", "--upsert", "--batch", "--from", knownHostsPath}
```
Comment (sbose78) [inline test/e2e/fixture/certs/certs.go:47]: Cool!

File: test/remote/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,86 @@
+FROM golang:1.16.2 AS go
```
Comment (sbose78) [inline test/remote/Dockerfile:1]: Could you please add a comment mentioning why this is needed, and where this is consumed ?

File: test/remote/README.md [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+export NAMESPACE=argocd-e2e
+# If you installed Argo CD via Operator, set accordingly
+# export ARGOCD_E2E_NAME_PREFIX=argocd-cluster
+./test/remote/generate-permissions.sh | kubectl apply -f -
+```
+
+### Build the repository container image
+
+You will need to build & publish a container image that will hold the required
+testing repositories.
```
Comment (sbose78) [inline test/remote/README.md:34]: Do you mean, code repositories?

File: test/remote/README.md [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+make manifests > /tmp/e2e-repositories.yaml
+```
+
+If you do not have kustomize installed, you need to manually edit the manifests
+at `test/remote/manifests/e2e_repositories.yaml` to point to the correct image.
+
+### Deploy the test container and additional permissions
+
+**Note:** The test container requires to be run in privileged mode for now, due
+to some processes running as root (this may change some day...).
```
Comment (sbose78) [inline test/remote/README.md:58]: 👍 Important note.

File: test/remote/README.md [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+./test/remote/run-e2e-remote.sh make test-local
+```
+
+### On remote OpenShift cluster
+
+You should first scale down GitOps Operator, since it will revert changes made
+during the tests instantly:
+
+```shell
+oc -n openshift-operators scale deployment --replicas=0 gitops-operator
```
Comment (sbose78) [inline test/remote/README.md:142]: Suggested replacement for upstream:

/s/gitops-operator/argocd-operator :)

File: test/remote/README.md [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+* `ARGOCD_E2E_ADMIN_PASSWORD` - the admin user's password to use
+* `ARGOCD_E2E_TEST_TIMEOUT` - timeout for the complete test suite, specified as duration (e.g. `2h` or `1h30m`)
+* `ARGOCD_E2E_DEFAULT_TIMEOUT` - timeout in seconds for each context operation (e.g. sync). Default `30`.
+* `ARGOCD_E2E_NAMESPACE` - the namespace where Argo CD is running in for the tests
+* `ARGOCD_E2E_NAME_PREFIX` - if your Argo CD installation has a name prefix (e.g. installed by the Operator), specify it here
+
+Furthermore, you can skip various classes of tests by setting the following to true:
+
+```
+# GnuPG features not yet available with GitOps Operator
```
Comment (sbose78) [inline test/remote/README.md:194]: Probably good enough to mention that the following may be disabled optionally if support for the same is not available in your installation of ArgoCD.

File: test/remote/generate-permissions.sh [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+	SERVERSANAME="${ARGOCD_E2E_NAME_PREFIX}-argocd-server"
+else
+	CRNAME="argocd-application-controller"
+	CRBNAME="argocd-application-controller"
+	CONTROLLERSANAME="argocd-application-controller"
+	SERVERSANAME="argocd-server"
+fi
+
+sed \
+	-e "s/##CRNAME##/${CRNAME}/g" \
```
Comment (sbose78) [inline test/remote/generate-permissions.sh:26]: Fancy ;)

**Review by sbose78 — COMMENTED** [review]
Will review further, but this work looks very sophisticated :)

File: test/remote/Dockerfile (hunk 3 above)
Comment (sbose78) [reply test/remote/Dockerfile:1]: Read further, I understand why :)

File: util/git/client.go [hunk 9]
```diff
@@ -503,7 +503,7 @@ func (m *nativeGitClient) VerifyCommitSignature(revision string) (string, error)
 // runWrapper runs a custom command with all the semantics of running the Git client
 func (m *nativeGitClient) runGnuPGWrapper(wrapper string, args ...string) (string, error) {
 	cmd := exec.Command(wrapper, args...)
-	cmd.Env = append(cmd.Env, fmt.Sprintf("GNUPGHOME=%s", common.GetGnuPGHomePath()), "LANG=C")
+	cmd.Env = append(cmd.Env, fmt.Sprintf("GNUPGHOME=%s", common.GetGnuPGHomePath()), "LANG=c")
```
Comment (alexmt) [inline util/git/client.go:506]: Can you please explain why this change is required?

File: test/remote/manifests/kustomization.yaml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+resources:
+- e2e-repositories.yaml
+
+images:
+- name: quay.io/jfischer-redhat/argocd-e2e-cluster
```
Comment (alexmt) [inline test/remote/manifests/kustomization.yaml:8]: image name is supposed to be overridden right? In this case let avoid using personal repo name here. May be rename to just `argocd-e2e-cluster` ?

File: test/remote/manifests/e2e-repositories.yaml [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    matchLabels:
+      app.kubernetes.io/name: argocd-e2e-server
+  template:
+    metadata:
+      labels:
+        app.kubernetes.io/name: argocd-e2e-server
+    spec:
+      containers:
+      - name: argocd-e2e-server
+        image: quay.io/jfischer-redhat/argocd-e2e-cluster:latest
```
Comment (alexmt) [inline test/remote/manifests/e2e-repositories.yaml:19]: I think we should avoid using personal repo name here.

**Review by alexmt — CHANGES_REQUESTED** [review]
Awesome work @jannfis . Added couple comments about using personal docker repo name.

**Comment by alexmt** [comment]
I'm sure this PR required a lot of effort. I think we should try to use it a much as we can to take advantage of it.

May be run every night agains https://cd.apps.argoproj.io/ 

@sbose78 , @jannfis do you plan to run it internally as well?

**Comment by sbose78** [comment]
> May be run every night agains https://cd.apps.argoproj.io/

+1 we should do this. Does this get deployed nightly ?

> do you plan to run it internally as well?


@alexmt Not internally per se,  rather use existing automation ( [OpenShift CI ](https://docs.ci.openshift.org/docs/) )  for spinning up multi-node clusters on AWS per PR with ease to test upstream Argo CD with public test logs (  [example](https://prow.ci.openshift.org/job-history/gs/origin-ci-test/pr-logs/directory/pull-ci-redhat-developer-gitops-operator-master-v4.7-e2e) ).

Generally speaking, the goal of running e2e tests against a deployed instance of Argo CD is : 
* Remove the need for manual testing prior to an Argo CD release
* Run this on merge-to-master ( and eventually make it a PR ) in an openshift-agnostic way.
* In the mid-term, the aim is to organize tests in such a way that their successful execution gives us a high-level of confidence 'as a user'. 

Initially, this is what we plan on doing:
1. Build   https://github.com/argoproj/argo-cd/ `master`  ( or use an existing built image )
2. Deploy a multi-node OpenShift on AWS
3. Deploy the freshly baked ArgoCD on OpenShift
4. Run e2e tests against the live instance.

All logs would be public therefore we should be able to catch failures and log bugs against them on this repository early on and potentially fix them. 

Once this process is stable enough, I would propose (1) to (4) to be done on every merge-to-master on https://github.com/argoproj/argo-cd/. 

Note,
* Given that this would be an additional webhook to Argo CD repository, this would have to go through the enhancement proposal process.
* No downstream-specific concepts would leak into the tests.

File: test/remote/README.md (hunk 4 above)
Comment (jannfis) [reply test/remote/README.md:34]: No, the repositories hold the manifests used during testing. But if that is misleading, we could rephrase it to "the required repositories that hold our test manifests" - would that be better understandable?

File: util/git/client.go (hunk 9 above)
Comment (jannfis) [reply util/git/client.go:506]: Oops. That is a left-over from testing, not a required change, and was committed by accident. I will revert it. Thanks for catching :+1:

File: test/remote/README.md (hunk 7 above)
Comment (jannfis) [reply test/remote/README.md:194]: Agreed. Rephrased sentence.

File: test/remote/manifests/kustomization.yaml (hunk 10 above)
Comment (jannfis) [reply test/remote/manifests/kustomization.yaml:8]: Yes, good hint. I have changed it to be just `argocd-e2e-cluster`. Thanks!

File: test/remote/manifests/e2e-repositories.yaml (hunk 11 above)
Comment (jannfis) [reply test/remote/manifests/e2e-repositories.yaml:19]: Changed.

File: test/e2e/fixture/certs/certs.go (hunk 1 above)
Comment (jannfis) [reply test/e2e/fixture/certs/certs.go:21]: Done

File: test/remote/README.md (hunk 6 above)
Comment (jannfis) [reply test/remote/README.md]: Done. Thanks!

**Comment by jannfis** [comment]
Thanks for review @sbose78 and @alexmt 

I have addressed all your comments, PTAL.

File: util/gpg/gpg.go [hunk 12]
```diff
@@ -149,7 +149,7 @@ const MaxVerificationLinesToParse = 40
 
 // Helper function to append GNUPGHOME for a command execution environment
 func getGPGEnviron() []string {
-	return append(os.Environ(), fmt.Sprintf("GNUPGHOME=%s", common.GetGnuPGHomePath()), "LANG=C")
+	return append(os.Environ(), fmt.Sprintf("GNUPGHOME=%s", common.GetGnuPGHomePath()), "LANG=c")
```
Comment (alexmt) [inline util/gpg/gpg.go:152]: I think this should be reverted as well.

**Review by alexmt — APPROVED** [review]
LGTM.

Just one more comment https://github.com/argoproj/argo-cd/pull/5895#discussion_r608148372  Please feel free to merge after resolving it.

File: util/gpg/gpg.go (hunk 12 above)
Comment (jannfis) [reply util/gpg/gpg.go]: Yes, indeed. Good catch again. Sorry for being a little inattentive here
